lib/charliecloud.py

```diff
@@ -291,9 +263,98 @@ class defined in this module."""
             layers[lh] = TT(fp, members)
          else:
             empty_cnt += 1
-      DEBUG("skipped %d empty layers" % empty_cnt)
+      VERBOSE("skipped %d empty layers" % empty_cnt)
       return layers
 
+   def metadata_init(self):
+      "Initialize empty metadata structure."
+      # Elsewhere can assume the existence and types of everything here.
+      self.metadata = { "arch": None,
+                        "cwd": "/",
+                        "env": dict(),
+                        "labels": dict(),
+                        "shell": ["/bin/sh", "-c"],
+                        "volumes": list() }  # set isn't JSON-serializable
+
+   def metadata_load(self):
+      """Load metadata file, replacing the existing metadata object. If
+         metadata doesn't exist, warn and use defaults."""
+      path = self.metadata_path // "metadata.json"
+      if (not path.exists()):
+         WARNING("no metadata for image; using defaults")
```

Comment:
Perhaps we should specify the image reference in this warning? Instead of just saying "image".